File: src/chivalry/stages.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { createPlan } from '../lib/stages.js';
import { installGlobals } from '../test-support/uo.js';
import { STAGES } from './config.js';

// The shipped Chivalry table, as opposed to src/lib/stages.test.ts which is about the mechanism

beforeEach(() => {
  installGlobals();
});

describe('STAGES', () => {
  it('is already in ascending order, so the plan is the table as written', () => {
    expect(createPlan(STAGES).stages.map((stage) => stage.upTo)).toEqual(
      STAGES.map((stage) => stage.upTo),
    );
  });

  // Bushido's table was written with `105` where 105.0 was meant - a band below every other one,
  // which therefore never ran. Any bound that is not a plausible skill value is that typo coming back.
  it('states every bound in the tenths getSkill reports', () => {
    for (const stage of STAGES) {
      expect(stage.upTo).toBeGreaterThanOrEqual(100);
      expect(stage.upTo).toBeLessThanOrEqual(1200);
    }
  });

  it('aims at 120.0, which needs the power scrolls', () => {
    expect(createPlan(STAGES).goal).toBe(1200);
  });

  it('charges mana for every band, so the mana wait always has a figure', () => {
    for (const stage of STAGES) {
      expect(stage.mana).toBeGreaterThan(0);
    }
  });

  // A row whose buff belongs to a different spell would gate the wrong thing. Only asserted of the
  // rows that carry one - Holy Light and Noble Sacrifice put nothing up.
  it('names the same spell in spell and buff on every row that has one', () => {
    for (const stage of STAGES) {
      if (stage.buff === undefined) {
        continue;
      }

      expect(BuffDebuffs[stage.buff]).toBe(Spells[stage.spell]);
    }
  });

  it('leaves the two instant spells without a buff, because the client publishes none', () => {
    const instant = STAGES.filter(
      (stage) => stage.spell === Spells.HolyLight || stage.spell === Spells.NobleSacrifice,
    );

    expect(instant).toHaveLength(2);

    for (const stage of instant) {
      expect(stage.buff).toBeUndefined();
    }
  });

  // Nothing a paladin casts here wants a cursor, and one opened by a run that never answers it breaks
  // every action after it
  it('wants a target cursor for nothing', () => {
    for (const stage of STAGES) {
      expect(stage.target).toBeUndefined();
    }
  });

  // The shard's own minimum for each spell, which a band that opened below would train nothing but
  // refusals against. The bound before each row is where that row starts.
  it('opens every band above the shard s minimum for its spell', () => {
    const minimum = new Map<Spells, number>([
      [Spells.ConsecrateWeapon, 150],
      [Spells.DivineFury, 250],
      [Spells.EnemyOfOne, 450],
      [Spells.HolyLight, 550],
      [Spells.NobleSacrifice, 650],
    ]);

    const plan = createPlan(STAGES);

    plan.stages.forEach((stage, index) => {
      const opensAt = index === 0 ? 400 : plan.stages[index - 1].upTo;

      expect(opensAt).toBeGreaterThanOrEqual(minimum.get(stage.spell) ?? 0);
    });
  });
});

describe('stageNow', () => {
  it('works the bands the table was written for', () => {
    const { stageNow } = createPlan(STAGES);

    expect(stageNow(400)?.spell).toBe(Spells.ConsecrateWeapon);
    expect(stageNow(450)?.spell).toBe(Spells.DivineFury);
    expect(stageNow(600)?.spell).toBe(Spells.EnemyOfOne);
    expect(stageNow(700)?.spell).toBe(Spells.HolyLight);
    expect(stageNow(900)?.spell).toBe(Spells.NobleSacrifice);
  });

  it('is finished at 120.0', () => {
    expect(createPlan(STAGES).stageNow(1200)).toBeUndefined();
  });
});

File: legion/src/chivalry/stages_test.py
import unittest

from chivalry.config import FIRST_BAND, STAGES
from uo.stages import goal_of, make_plan, stage_now

# The shard's own minimum for each spell
MINIMUM = {
    "Consecrate Weapon": 15.0,
    "Divine Fury": 25.0,
    "Enemy of One": 45.0,
    "Holy Light": 55.0,
    "Noble Sacrifice": 65.0,
}


class ShippedStagesTest(unittest.TestCase):
    def setUp(self):
        self.plan = make_plan(STAGES)

    def test_the_table_is_already_in_the_order_it_trains_in(self):
        self.assertEqual(STAGES, self.plan)

    def test_the_last_band_is_where_a_power_scroll_leaves_off(self):
        self.assertEqual(goal_of(self.plan), 120.0)

    def test_every_band_costs_mana_and_tithing(self):
        for stage in STAGES:
            self.assertGreater(stage["mana"], 0, stage["spell"])
            self.assertGreater(stage["tithing"], 0, stage["spell"])

    def test_every_band_paces_itself(self):
        for stage in STAGES:
            self.assertIn("cast_timeout", stage, stage["spell"])
            self.assertIn("cast_delay", stage, stage["spell"])

    # Nothing a paladin casts here raises a cursor, and one left open breaks every action after it
    def test_no_band_answers_a_cursor(self):
        for stage in STAGES:
            self.assertNotIn("target", stage, stage["spell"])

    def test_only_the_three_self_buffs_prove_themselves_by_the_bar(self):
        buffed = [stage for stage in STAGES if "buff" in stage]

        self.assertEqual([stage["spell"] for stage in buffed],
                         ["Consecrate Weapon", "Divine Fury", "Enemy of One"])

        self.assertEqual([stage["buff"] for stage in buffed],
                         ["ConsecrateWeapon", "DivineFury", "EnemyOfOne"])

        for stage in buffed:
            self.assertEqual(stage["title"], stage["spell"])

    def test_only_consecrate_weapon_wants_a_weapon(self):
        self.assertEqual([stage["spell"] for stage in STAGES if stage.get("needs_weapon")],
                         ["Consecrate Weapon"])

    def test_every_band_opens_above_the_shards_minimum_for_its_spell(self):
        opens_at = FIRST_BAND

        for stage in self.plan:
            self.assertGreaterEqual(opens_at, MINIMUM[stage["spell"]], stage["spell"])
            opens_at = stage["up_to"]

    def test_hands_over_on_the_exclusive_edge(self):
        self.assertEqual(stage_now(self.plan, 44.9)["spell"], "Consecrate Weapon")
        self.assertEqual(stage_now(self.plan, 45.0)["spell"], "Divine Fury")
        self.assertEqual(stage_now(self.plan, 60.0)["spell"], "Enemy of One")
        self.assertEqual(stage_now(self.plan, 70.0)["spell"], "Holy Light")
        self.assertEqual(stage_now(self.plan, 90.0)["spell"], "Noble Sacrifice")
        self.assertEqual(stage_now(self.plan, 119.9)["spell"], "Noble Sacrifice")

    def test_is_finished_at_the_goal(self):
        self.assertIsNone(stage_now(self.plan, 120.0))
